pkg/connection/store/external/store.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package external implements a gRPC client for external secret store plugins.
+package external
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	ess "github.com/crossplane/crossplane-runtime/apis/proto/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/connection/store"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+)
+
+// Error strings.
+const (
+	errLoadCert  = "cannot load client certificates"
+	errLoadCA    = "cannot load CA certificate"
+	errInvalidCA = "invalid CA certificate"
+
+	errGet    = "cannot get secret"
+	errApply  = "cannot apply secret"
+	errDelete = "cannot delete secret"
+)
+
+var (
+	certsPathFmt = "/certs/%s"
+	caCertFile   = "ca.crt"
+	tlsCertFile  = "tls.crt"
+	tlsKeyFile   = "tls.key"
+)
+
+// SecretStore is an External Secret Store.
+type SecretStore struct {
+	client     ess.ExternalSecretStoreServiceClient
+	kubeClient client.Client
+	ctx        context.Context
+	config     *v1.Config
+}
+
+// loadKeyPair loads CA and client certificates.
+func loadKeyPair() (credentials.TransportCredentials, error) {
+	certificate, err := tls.LoadX509KeyPair(fmt.Sprintf(certsPathFmt, tlsCertFile), fmt.Sprintf(certsPathFmt, tlsKeyFile))
+	if err != nil {
+		return nil, errors.Wrap(err, errLoadCert)
+	}
+
+	ca, err := os.ReadFile(fmt.Sprintf(certsPathFmt, caCertFile))
+	if err != nil {
+		return nil, errors.Wrap(err, errLoadCA)
+	}
+
+	pool := x509.NewCertPool()
+	if !pool.AppendCertsFromPEM(ca) {
+		return nil, errors.New(errInvalidCA)
+	}
+
+	tlsConfig := &tls.Config{
+		MinVersion:   tls.VersionTLS12,
+		Certificates: []tls.Certificate{certificate},
+		RootCAs:      pool,
+	}
+
+	return credentials.NewTLS(tlsConfig), nil
+}
+
+// NewSecretStore returns a new External SecretStore.
+func NewSecretStore(ctx context.Context, kube client.Client, cfg v1.SecretStoreConfig) (*SecretStore, error) {
+	kp, err := loadKeyPair()
```

Comment:
I assume this client will use the same certs no matter which plugin server it talks to (when there are more than once). Should we move `loadKeyPair` to a previous stage (e.g., startup) so that we wouldn't need to do it each time we want to build a secret store, which is expected to happen in each reconcile?